### ISSUE TITLE
PyCall segmentation fault when running a pure Julia task in parallel

### ISSUE THREAD

**Issue by Deduction42** [JuliaLang/julia#33893]
I'm not sure if this is a package issue or an issue with the new multi-threading tool with Julia 1.3-rc5. I am trying to use Python's fastparquet to write a dataframe into a parquet file (as there is no Julia package that allows me to do this). While doing this, I am running a task in parallel to process a Julia Dataframe. However, trying to do so results in a segfault. I know that I probably can't run multiple PyCalls with multithreading (as Python is not thread-safe), but does this mean I also cannot do something in pure Julia while making a Python call? I'm not sure if this is a bug or  a system limitation (which should be documented somewhere but I can't find it).

```
signal (11): Segmentation fault
in expression starting at /home/********/Desktop/Parquet Append Testing/CsvProcessing.jl:176
_PyObject_Malloc at /home/********/miniconda3/lib/libpython3.7m.so.1.0 (unknown line)
PyUnicode_New at /home/********/miniconda3/lib/libpython3.7m.so.1.0 (unknown line)
_PyUnicodeWriter_PrepareInternal at /home/********/miniconda3/lib/libpython3.7m.so.1.0 (unknown line)
PyUnicode_DecodeUTF8Stateful at /home/********/miniconda3/lib/libpython3.7m.so.1.0 (unknown line)
PyObject at /home/********/.julia/packages/PyCall/ttONZ/src/exception.jl:81
array2py at /home/********/.julia/packages/PyCall/ttONZ/src/conversions.jl:314
array2py at /home/********/.julia/packages/PyCall/ttONZ/src/conversions.jl:319
array2py at /home/********/.julia/packages/PyCall/ttONZ/src/conversions.jl:328 [inlined]
PyObject at /home/********/.julia/packages/PyCall/ttONZ/src/conversions.jl:330 [inlined]
_pycall! at /home/********/.julia/packages/PyCall/ttONZ/src/pyfncall.jl:24
_pycall! at /home/********/.julia/packages/PyCall/ttONZ/src/pyfncall.jl:11
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2135 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
#pycall#109 at /home/********/.julia/packages/PyCall/ttONZ/src/pyfncall.jl:83
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2135 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1631 [inlined]
jl_f__apply at /buildworker/worker/package_linux64/build/src/builtins.c:627
pycall at /home/********/.julia/packages/PyCall/ttONZ/src/pyfncall.jl:83
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2135 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
setindex! at /home/********/.julia/packages/Pandas/wBpaf/src/Pandas.jl:183
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2135 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
df2pandas at /home/********/Desktop/Parquet Append Testing/CsvProcessing.jl:40
macro expansion at ./util.jl:155 [inlined]
write_parquet at /home/********/Desktop/Parquet Append Testing/CsvProcessing.jl:48
write_parquet at /home/********/Desktop/Parquet Append Testing/CsvProcessing.jl:54 [inlined]
read_write at /home/********/Desktop/Parquet Append Testing/CsvProcessing.jl:131
macro expansion at ./util.jl:155 [inlined]
run_parser at /home/********/Desktop/Parquet Append Testing/CsvProcessing.jl:169
unknown function (ip: 0x7fe8cf52d71c)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2141 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1631 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:328
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:417
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:368 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:778
jl_interpret_toplevel_thunk_callback at /buildworker/worker/package_linux64/build/src/interpreter.c:888
unknown function (ip: 0xfffffffffffffffe)
unknown function (ip: 0x7fe84161da8f)
unknown function (ip: 0xffffffffffffffff)
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:897
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:814
jl_parse_eval_all at /buildworker/worker/package_linux64/build/src/ast.c:873
include_string at ./loading.jl:1075
#139 at /home/********/.julia/packages/Atom/lBERI/src/eval.jl:138
withpath at /home/********/.julia/packages/CodeTools/xGemk/src/utils.jl:30
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2141 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
withpath at /home/********/.julia/packages/Atom/lBERI/src/eval.jl:47
#138 at /home/********/.julia/packages/Atom/lBERI/src/eval.jl:135 [inlined]
with_logstate at ./logging.jl:395
with_logger at ./logging.jl:491 [inlined]
#137 at /home/********/.julia/packages/Atom/lBERI/src/eval.jl:134 [inlined]
hideprompt at /home/********/.julia/packages/Atom/lBERI/src/repl.jl:85
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2141 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
macro expansion at /home/********/.julia/packages/Atom/lBERI/src/eval.jl:133 [inlined]
macro expansion at /home/********/.julia/packages/Media/ItEPc/src/dynamic.jl:24 [inlined]
#136 at /home/********/.julia/packages/Atom/lBERI/src/eval.jl:122
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2141 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
handlemsg at /home/********/.julia/packages/Atom/lBERI/src/comm.jl:164
unknown function (ip: 0x7fe8d2e1df3b)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2135 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1631 [inlined]
jl_f__apply at /buildworker/worker/package_linux64/build/src/builtins.c:627
#19 at ./task.jl:333
unknown function (ip: 0x7fe8d2e1cc47)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2135 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2305
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1631 [inlined]
start_task at /buildworker/worker/package_linux64/build/src/task.c:659
unknown function (ip: 0xffffffffffffffff)
Allocations: 288037862 (Pool: 287995801; Big: 42061); GC: 105
```

If you have a question please search or post to our Discourse site: https://discourse.julialang.org.
We use the GitHub issue tracker for bug reports and feature requests only.

If you're submitting a bug report, be sure to include as much relevant information as
possible, including a minimal reproducible example and the output of `versioninfo()`.
If you're experiencing a problem with a particular package, open an issue on that
package's repository instead.

Thanks for contributing to the Julia project!

**Comment by stevengj** [comment]
I would hope that as long as you only ever use Python from a single thread you should be okay … can you put together a minimal example that illustrates the problem? ( i.e. delete every line in your code until the problem goes away...)

**Comment by mdav2** [comment]
Bit late to the party here, but I ran into a very similar issue and was able to resolve it. I don't understand why, but moving all of my PyCall code (like `using PyCall` and `pyimport` satements) into one file did the trick. 

In my case I am writing my own (Julia) package that uses a python package to obtain some methods that are difficult to code. Trouble was, I needed those methods both inside my package _and_ when using my package. So I had two `using PyCall` regions. Coalescing them and exporting the python package from my package, then using that while using my package resolved the issue. 

```julia
module dummy
using PyCall
const psi4 = PyNULL()
function __init__()
     copy!(psi4, pyimport("psi4")
end
end 
export psi4
```

**Comment by Deduction42** [comment]
I managed to figure out what the issue was. I have a function that reads a CSV file into a dataframe and does some transformations: 

`process_file (FileName::String)`

I use this function on a list of CSV files assigned to different tasks

`function working_dir_tasks(WorkingDir::String, process_file::Function; MaxFiles=(Threads.nthreads-1) )

    #Read the directory
    FileList = get_working_files()
    FileList = FileList[1:min(MaxFiles,length(FileList))]
    TaskList = Task[]

    for File in FileList
        t = Threads.@spawn process_file(File)
        push!(TaskList, t)
    end

    return TaskList
end`

In the main program, I have a loop. If I try to write a list of tables using pure Julia code WHILE preprocessing other files, it crashes pretty reliably.
`    while true
            TaskList = working_dir_tasks(WorkingDir, custom_processing; MaxFiles=nProcess )

            #Write previous list of tables while getting new ones in background
            @time for Table in TableList
                write_parquet(ParquetFile, Table)
            end

            #Get new list from background tasks
            @time TableList = fetch.(TaskList)
    end`

If, however I fetch the results before running the Python code, it doesn't crash
`    while true
            TaskList = working_dir_tasks(WorkingDir, custom_processing; MaxFiles=nProcess )

            #Get new list from background tasks
            @time TableList = fetch.(TaskList)

            #Write previous list of tables while getting new ones in background
            @time for Table in TableList
                write_parquet(ParquetFile, Table)
            end
    end`

**Comment by Deduction42** [comment]
It would also be interesting to note that the only thing I'm using Python for here is writing files to parquet. This code is literally trying to write parquet files using Python while reading and processing files in Julia. If Julia had a functional Parquet writer, none of this mess would have to happen.

**Comment by stevengj** [comment]
>  It fails when trying to convert a Julia dataframe to a Pandas dataframe.

Doesn't this involve calling Python code, which is not thread-safe?

**Comment by Deduction42** [comment]
Yes, but I'm only ever running the Python code on one thread. I'm running Julia on the other threads. However, I keep getting memory conflicts. I've tried a number of different configurations just to see, and this is what I found:

(1) Try to run pure Julia threads while calling Python to convert a Julia dataframe to pandas 
     Result: Crashes quickly
(2) Don't run julia threads while calling python to convert to pandas, but run Julia threads again while calling python fastparquet to write pandas to disk parquet:
     Result: Crashes eventually (but runs for a while)
(3) Try to do something completely unrelated in Python like plotting random data:
     Result: I haven't seen it crash yet
(4) Don't run any julia threads while calling python
     Result: Ran for 48 hours without crashing

So it looks like we really can't run Julia on other threads while calling python on the main thread.
